kernel_remove: init script to remove a specified kernel

thefossguy-ciq opened the pull request:
This script basically cleans up all the possible paths that might be populated from a `make modules_install` and `make install`. Given that the kernel is removed, the grub entry is also removed with the help of `grubby`.

This is an ad-hoc script created to quickly remove a kernel without much fuss about it.

---

The script only accepts the kernel version and is used like following:

```
$ ./kernel_remove.sh $KERNEL_VERSION

$ ./kernel_remove.sh 4.18.0-ciqlts8_8-6a8fa3030700+
```

And that's all. It will remove all installed files and even remove the grub entry.

PlaidCat commented:
Could you add a check for making sure you're not deleting the running kernel?

Additionally could you add an Entry to the ReadMe explaining this and the expected use case and limitations?

Otherwise thanks this will be helpful for full systems or those that don't treat their VM's as disposable (i know this is important to some)

thefossguy-ciq committed:
kernel_remove: init script to remove a specified kernel

This script basically cleans up all the possible paths that might be
populated from a `make modules_install` and `make install`. Given that
the kernel is removed, the grub entry is also removed with the help of
`grubby`.

This is an ad-hoc script created to quickly remove a kernel without much
fuss about it.

thefossguy-ciq commented:
Some extra documentation that I am not sure goes in the README.

### Trying to remove a kernel installed as an RPM
```bash
$ uname -r
4.18.0-ciqlts8_6-83208b00c+

$ ls -alh /lib/modules/
total 16K
drwxr-xr-x.  6 root root  161 Jul 14 21:56 .
dr-xr-xr-x. 40 root root 4.0K Jul 14 17:59 ..
drwxr-xr-x.  6 root root 4.0K Jul 14 21:57 4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64
drwxr-xr-x.  6 root root 4.0K Jul 14 21:57 4.18.0-372.32.1.el8_6.x86_64
drwxr-xr-x.  3 root root   19 Jul 14 17:42 4.18.0-372.9.1.el8.x86_64
drwxr-xr-x.  3 root root 4.0K Jul 14 20:56 4.18.0-ciqlts8_6-83208b00c+

$ ./kernel_remove.sh 4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64
+ requested_kernel_to_remove=4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64
++ dnf repoquery --latest-limit=1 --queryformat '%{version}-%{release}' kernel-core
+ latest_rpm_kernel=4.18.0-372.32.1.el8_6.86ciq_lts.12.1
++ uname -r
+ current_kernel=4.18.0-ciqlts8_6-83208b00c+
+ '[' -z 4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64 ']'
+ '[' 4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64 == 4.18.0-372.32.1.el8_6.86ciq_lts.12.1 ']'
++ uname -m
+ '[' 4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64 == 4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64 ']'
+ echo 'ERROR: Will not remove the *latest kernel* installed as an RPM package'
ERROR: Will not remove the *latest kernel* installed as an RPM package
+ exit 1
```

### Trying to remove the current kernel
```bash
$ uname -r
4.18.0-ciqlts8_6-83208b00c+

$ ./kernel_remove.sh 4.18.0-ciqlts8_6-83208b00c+
+ requested_kernel_to_remove=4.18.0-ciqlts8_6-83208b00c+
++ dnf repoquery --latest-limit=1 --queryformat '%{version}-%{release}' kernel-core
+ latest_rpm_kernel=4.18.0-372.32.1.el8_6.86ciq_lts.12.1
++ uname -r
+ current_kernel=4.18.0-ciqlts8_6-83208b00c+
+ '[' -z 4.18.0-ciqlts8_6-83208b00c+ ']'
+ '[' 4.18.0-ciqlts8_6-83208b00c+ == 4.18.0-372.32.1.el8_6.86ciq_lts.12.1 ']'
++ uname -m
+ '[' 4.18.0-ciqlts8_6-83208b00c+ == 4.18.0-372.32.1.el8_6.86ciq_lts.12.1.x86_64 ']'
+ '[' 4.18.0-ciqlts8_6-83208b00c+ == 4.18.0-ciqlts8_6-83208b00c+ ']'
+ echo 'ERROR: Will not remove the current kernel'
ERROR: Will not remove the current kernel
+ exit 1
```

thefossguy-ciq commented on kernel_remove.sh:
```diff
@@ -0,0 +1,30 @@
+    [ -d "/boot/dtb-${requested_kernel_to_remove}" ] && sudo rm -fr "/boot/dtb-${requested_kernel_to_remove}"
+    [ -f "/boot/config-${requested_kernel_to_remove}" ] && sudo rm -f "/boot/config-${requested_kernel_to_remove}"
+    [ -f "/boot/System-${requested_kernel_to_remove}" ] && sudo rm -f "/boot/System-${requested_kernel_to_remove}"
+    [ -f "/boot/vmlinuz-${requested_kernel_to_remove}" ] && sudo rm -f "/boot/vmlinuz-${requested_kernel_to_remove}"
+    # If $requested_kernel_to_remove is current default kernel, grubby will make the next-in-line kernel default
+    sudo grubby --remove-kernel="/boot/vmlinuz-${requested_kernel_to_remove}"
```
No it doesn't. `grubby` interacts with the GRUB2 bootloader, not files on the filesystem.

PlaidCat commented on kernel_remove.sh:
```diff
@@ -0,0 +1,33 @@
+set -xeuf -o pipefail
+
+requested_kernel_to_remove="${1:-}"
+latest_rpm_kernel="$(rpm -q kernel-core --qf "%{VERSION}-%{RELEASE}\n" 2>/dev/null | sort --version-sort | tail -1)"
```
I like this maybe a fast follow improvement is to protect any `dnf` installed kernel.

Rather than the simple check below it would turn into a loop but this is more than what I asked for so thats good.

PlaidCat reviewed (dismissed):
Comment is non blocking 

:shipit:

thefossguy-ciq commented:
Latest force push prevents **any** kernel installed as an RPM package (instead of only the latest one).

PlaidCat reviewed (approved):
Thanks for the update it wasn't required but good to see

:shipit:
